src/macaron/artifact/local_artifact.py

```diff
@@ -0,0 +1,275 @@
+    >>> construct_local_artifact_paths_glob_pattern_maven_purl(purl)
+    ['com/oracle/macaron/macaron/0.13.0']
+    """
+    if not maven_purl.type == "maven":
+        return None
+
+    group = maven_purl.namespace
+    artifact = maven_purl.name
+    version = maven_purl.version
+
+    if group is None or version is None:
+        return None
+
+    return [construct_maven_repository_path(group, artifact, version)]
+
+
+def construct_local_artifact_paths_glob_pattern_pypi_purl(pypi_purl: PackageURL) -> list[str] | None:
+    """Return a list of glob pattern(s) representing python artifacts in a virtual environment.
+
+    The glob pattern(s) can be used to search in `<...>/<python_venv>/lib/python3.x/site-packages`
+    directory.
+
+    Parameters
+    ----------
+    maven_purl : PackageURL
+        A pypi type PackageURL instance.
+
+    Returns
+    -------
+    list[str] | None
+        A list of glob patterns or None if an error happened.
+
+    Examples
+    --------
+    >>> from packageurl import PackageURL
+    >>> purl = PackageURL.from_string("pkg:pypi/django@1.11.1")
+    >>> construct_local_artifact_paths_glob_pattern_pypi_purl(purl)
+    ['django', 'django-1.11.1.dist-info', 'django-1.11.1.data']
+    """
+    if not pypi_purl.type == "pypi":
+        return None
+
+    name = pypi_purl.name
+    version = pypi_purl.version
+
+    if version is None:
+        return None
+
+    # These patterns are from the content of a wheel file, which are extracted into the site-packages
+    # directory. References:
+    # https://packaging.python.org/en/latest/specifications/binary-distribution-format/#file-contents
+    glob_patterns = []
+    glob_patterns.append(name)
+    glob_patterns.append(f"{name}-{version}.dist-info")
+    glob_patterns.append(f"{name}-{version}.data")
+
+    return glob_patterns
+
+
+def find_artifact_paths_from_local_maven_repo(
+    local_maven_repo: str,
+    glob_patterns: list[str],
+) -> list[str] | None:
+    """Find maven artifacts within a local maven repository directory.
+
+    ``local_maven_repo`` should be in format `<...>/.m2/repository`.
+
+    Parameters
+    ----------
+    local_maven_repo: str
+        The path to the directories to find artifacts.
+    glob_patterns: list[str]
+        The list of glob patterns that matches to artifact file names.
+
+    Returns
+    -------
+    list[str] | None
+        The list of path to found artifacts in the form of ``local_maven_repo``/<artifact_specific_path>
+    """
+    if not os.path.isdir(local_maven_repo):
+        return None
+
+    artifact_paths = []
+    for pattern in glob_patterns:
+        found_paths = glob.glob(
+            root_dir=local_maven_repo,
+            pathname=pattern,
+        )
+
+        for found_path in found_paths:
+            full_path = os.path.join(local_maven_repo, found_path)
+            if os.path.isdir(full_path):
+                artifact_paths.append(full_path)
+
+    return artifact_paths
```

Comment:
```suggestion
    return artifact_paths or None
```

Reply:
My original intention with this function is that:
- If it returns None, it means an error happened.
- If it returns a list, it means no error happened. Therefore, if an empty list is returned, everything went fine but we cannot find the local artifact for the input PURL. We shouldn't treat this as an error.
That's why at this line I don't return None because by doing so I could introduce a third behavior, which I don't think is ideal:
- A None value returned can indicate an error, or no local artifacts found.

I can think of 3 solutions:
1. I can use Exception for error cases, None value for when no local artifact is found, and a non-empty list to indicate some local artifacts are found. But I'm not sure if this is desired.
2. Use Exception instead of None to indicate error and remove the `None` usages in this function.
3. I can mention my original intention more clearly in the doc strings.

Please let me know what you think.

Reply:
I think with how Python is, checking that something is actually None and not just falsey is unusual. If there is actually a case where we want to detect the error, maybe an exception is better.

I only really suggested this change because it enables the small simplification I suggested for the `_get_local_artifact_path_for_build_tool_purl_type` function. I don't think either change is essential.

Reply:
I agree. I will switch to use Exception for this one instead. Thanks.

Reply:
Addressed in https://github.com/oracle/macaron/pull/864/commits/7adbee2cbd4ee972af55aefc50f7acd4fe6214c2